include/core/shape.h

```diff
@@ -133,6 +134,950 @@ class Shape : public AutoDerivativeFunction<dim>
+                if (new_distance > current_distance - precision * precision)
+                  {
+                    // Initialize the container for the value.
+                    std::vector<double> diff_results;
+                    diff_results.resize(search_direction.size());
+                    Point<dim> best_point;
+                    double     best_dist = current_distance;
+                    for (unsigned int d = 0; d < search_direction.size(); ++d)
+                      {
+                        Tensor<1, dim> perturbation =
+                          search_direction[d] * max_step;
+                        value_first_component =
+                          this->value(current_point + perturbation);
+                        value_second_component =
+                          shape.value(current_point + perturbation);
+                        new_distance = smooth_max(value_first_component,
+                                                  value_second_component);
+                        diff_results[d] =
+                          (new_distance - current_distance) / max_step;
+                        // Check if the Cartesian search is better than the last
+                        // candidate.
+                        if (new_distance < best_dist - precision * precision)
+                          {
+                            best_dist  = new_distance;
+                            best_point = current_point + perturbation;
+                          }
+                      }
+                    // Store the current results.
+                    if (best_dist < current_distance - precision * precision)
+                      {
+                        current_distance = best_dist;
+                        current_point    = best_point;
+                      }
+                    Point<dim> extra_guess;
+                    // If none of the cartesian candidates are better, we use
+                    // the value to find the point that minimizes the numerical
+                    // gradient we evaluated with the cartesian guess.
+                    if (current_distance >
+                        previous_value - precision * precision)
+                      {
+                        // Define the guess point.
+                        if constexpr (dim == 2)
+                          {
+                            new_point[0] =
+                              previous_position[0] -
+                              ((diff_results[0] - diff_results[1]) / 2) /
+                                ((diff_results[0] + diff_results[1]) /
+                                 max_step);
+                            new_point[1] =
+                              previous_position[1] -
+                              ((diff_results[2] - diff_results[3]) / 2) /
+                                ((diff_results[2] + diff_results[3]) /
+                                 max_step);
+                          }
```

Comment:
Same comment as before. INstead of having two blocks of code, do what is common for both dim values and put the additional stuff in a const expr dim==3 block